File: src/easy/Q0897.cpp
// 897. Increasing Order Search Tree
// https://leetcode.com/problems/increasing-order-search-tree/

#include <functional>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;

  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

// Time: O(n), Space: O(n), where n is the number of nodes in the tree.
auto increasing_bst(TreeNode *root) -> TreeNode * {
  auto vals = vector<int>{};

  function<void(TreeNode *)> dfs = [&dfs, &vals](TreeNode *node) {
    if (node == nullptr) {
      return;
    }
    dfs(node->left);
    vals.push_back(node->val);
    dfs(node->right);
  };

  dfs(root);
  auto dummy = new TreeNode();
  auto current = dummy;
  for (const int val : vals) {
    current->right = new TreeNode(val);
    current = current->right;
  }
  return dummy->right;
}

File: src/main/python/easy/Q0897.py
"""
897. Increasing Order Search Tree
https://leetcode.com/problems/increasing-order-search-tree/
"""

from __future__ import annotations
from dataclasses import dataclass


@dataclass
class TreeNode:
    val: int = 0
    left: TreeNode | None = None
    right: TreeNode | None = None


# Time: O(n), Space: O(n), where n is the number of nodes in the tree.
def increasing_bst(root: TreeNode) -> TreeNode:
    vals = []

    def dfs(node: TreeNode | None) -> None:
        if node is None:
            return
        dfs(node.left)
        vals.append(node.val)
        dfs(node.right)

    dfs(root)
    dummy = current = TreeNode()
    for val in vals:
        current.right = TreeNode(val)
        current = current.right
    return dummy.right
